### PR TITLE
accelerate the masked slic superpixels algorithm

### PR DESCRIPTION
**Pull request by grlee77**
## Description

This PR is related to #4902. For the image of shape (400, 400, 100) used in the benchmarks (`with multichannel=False`), I get a time of 59 seconds prior to this change where >60% of the time is taken up by the call to `kmeans2` in `_get_mask_centroids`. 

Two ways to make the kmeans2 call run faster are:
1.) use a subset of the full set of points for the first argument.
2.) lower the number of k-means iterations from SciPy's default value of 10.

This gives a reduction from 59.4 s to 19.9 s to run the `time_mask_slic` benchmark. Before this change ~60% of the execution time is in `_get_mask_slic` while afterwards it is around 5%.

After this change the `slic` output is still deterministic, but the starting centroids would not be identical to prior releases. If that is a problem, we could add a flag to only enable this acceleration on request.

One unresolved question is how many iterations and what density of samples is most appropriate. I set `dense_factor=10` empirically. It may be possible to use a somewhat lower value, but there is very little further performance gain, at least for the benchmark case. Also, with `dense_factor` of 1 or 2 I sometimes see warnings like:

```
scipy/cluster/vq.py:574:
UserWarning: One of the clusters is empty. Re-run kmeans with a different initialization.
  warnings.warn("One of the clusters is empty. "
```

For the `mask_time_slic` benchmark case there are 25,992,000 nonzero voxels in the mask, but after the change in this PR we will only use a much smaller subset of these in `kmeans2`. After this change, for the benchmark example, ``n_dense = (dense_factor**ndim) * n_centroids`` is only 100,000 out of 25,992,000 points. 

## Checklist

- [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- Gallery example in `./doc/examples` (new features only)
- Benchmark in `./benchmarks`, if your changes aren't covered by an
  existing benchmark
- Unit tests
- Clean style in [the spirit of PEP8](https://www.python.org/dev/peps/pep-0008/)

## For reviewers

- Check that the PR title is short, concise, and will make sense 1 year
  later.
- Check that new functions are imported in corresponding `__init__.py`.
- Check that new features, API changes, and deprecations are mentioned in
      `doc/release/release_dev.rst`.

### CODE REVIEW

**Comment by grlee77** [comment]
Results of the masked slic example before and after this PR. Runtime for the masked 2D case decreases from 545 ms to 303 ms, so less than the nearly factor 3 runtime decrease for the 3D benchmark, but still a substantial reduction.

mask_slic example on master:
mask_slic_master

vs. this PR (note similar qualitative superpixels in the masked case, but not the same ones as in current master)
mask_slic_pr

**Comment by emmanuelle** [comment]
Thanks, this is a very smart optimization :-). Could you please:
- add an item in the `release_dev.rst` (probably under Improvements)
- add a test so that the two cases of the new if/else block are covered by tests?
When I was profiling the module yesterday I thought maybe it would also be possible to gain a little my doing early stopping (small number of iterations) in this part of the code
```python
if update_centroids:
        # Step 2 of the algorithm [3]_
        _slic_cython(image, mask, segments, step, max_iter, spacing,
                     slic_zero, ignore_color=True,
                     start_label=start_label)
```
but I'm not so sure of what this part does (it looks to me like a simple kmeans if color is not taken into account), so not 100% sure.

**Comment by grlee77** [comment]
Actually, I think we can also get a large speedup by changing the SLIC search window area to match the size in the original publication. Specifically, do you know why the following code is +/- `2*step` and not just +/- `step`?

https://github.com/scikit-image/scikit-image/blob/66f78a4a3aeadc423f4d2389dce4b6666ea50a2f/skimage/segmentation/_slic.pyx#L141-L146

If you look at Figure 2 [in this paper by the original authors](https://infoscience.epfl.ch/record/177415/files/Superpixel_PAMI2011-2.pdf) you can see that the search window is size `2*step` per size, but our code is `4*step` per side. Changing that would make a factor of `2**ndim` reduction in the number of times the inner loop is called.

**Comment by grlee77** [comment]
> When I was profiling the module yesterday I thought maybe it would also be possible to gain a little my doing early stopping (small number of iterations) in this part of the code

Yes, I think we could do that. I was playing with it on an example and the centroids actually don't converge until many iterations (>150 for the `plot_mask_slic` example data), but we don't need to reach convergence since the real `slic` call below is going to modify the shapes substantially anyways. Also, we definitely shouldn't need a 4*step search window size (see above) for these initial iterations even if we choose to keep that for the main `slic` iterations below.

**Comment by emmanuelle** [comment]
About the size of the window, this is a very interested comment. Maybe the idea was to allow more convoluted superpixels, which follow gradients in a better way, but 
- if one wants to have elongated / convoluted superpixels, one should use felzenswalb instead (for example)
- a consequence of our generous window size is that superpixels are often not connected, and that there is a lot of work to do in the `enforce_connectivity` step. See for example below 
superpixels_before_connectivity

So I really think it's worth changing the window size to the one of the original article. Probably in an other PR since it will be a bit more controversial than this one.

**Comment by emmanuelle** [comment]
@grlee77 the PR is already a big improvement, we can leave further optimization (reducing the size of the search window as in the original article). If you have the time to add a line in the release notes it'd be awesome, otherwise I can do it in the "catching-up on release notes" PR.

**Review by emmanuelle — APPROVED** [review]
Thanks @grlee77 the PR is ready to go as far as I'm concerned! Now we need a second review :-).

**Comment by emmanuelle** [comment]
ping @scikit-image/core

**Review by jni — APPROVED** [review]

**Comment by jni** [comment]
Thank you @grlee77 and thank you @emmanuelle for the ping. =)

**Pull request merged**
